### ISSUE TITLE
OCLGRIND FATAL ERROR

### ISSUE THREAD

**Issue by lalanne** [jrprice/Oclgrind#69]
Hello (great tool btw!), 
I'm testing this tool specifically for getting the histogram of SPIR instructions, I installed that last version of the binaries (15.5). when i ran the tool with an Nbody OpenCL application (from the OpenDwarfs benchmark) I got this error.

OCLGRIND FATAL ERROR (/home/jprice/oclgrind-testing/cmake/src/oclgrind-15.5/src/core/common.cpp:142)

If you need more info i can compile the sources if its necessary. At the beginning the output is this:

Platform Chosen : University of Bristol
Number of available devices: 1 
Device Chosen : Oclgrind Simulator .

Ok, any help or info that you need please ping me.

Christian,

**Comment by jrprice** [comment]
Thanks for the bug report. I've managed to reproduce this, so will try and produce a fix shortly.

Out of interest, was the line you posted the only error output, or did it also dump out a bunch of info like the error below?

    OCLGRIND FATAL ERROR (/home/jprice/oclgrind-testing/cmake/src/oclgrind-15.5/src/core/common.cpp:142)
    Unsupported unsigned int size: 0 bytes
	    Kernel: calc_potential_single_step_dev
	    Entity: Global(0,6,0) Local(0,0,0) Group(0,6,0)
	      switch i2 %defs.sroa.4.0, label %141 [
	        i2 -1, label %109
	        i2 -2, label %124
	      ], !dbg !102
	    At line 168 of input.cl:
	      if (defs.region == 3)  /* Dist to surf > ion_excl_rad */


The issue is with one of the optimisation passes that LLVM is running before generating SPIR. The benchmark should run if you disable optimizations:

    oclgrind --build-options "-cl-opt-disable" ./gemnoui ...

This works for me, although the instruction histogram you get back will be less representative of what would be executed by a real device.

**Comment by lalanne** [comment]
Yes, this is the complete dump of errors

OCLGRIND FATAL ERROR (/home/jprice/oclgrind-testing/cmake/src/oclgrind-15.5/src/core/common.cpp:142)
Unsupported unsigned int size: 0 bytes
        Kernel: calc_potential_single_step_dev
        Entity: Global(0,0,0) Local(0,0,0) Group(0,0,0)
          switch i2 %defs.sroa.4.0, label %141 [
            i2 -1, label %109
            i2 -2, label %124
          ], !dbg !106
        At line 176 of input.cl:
          if (defs.region == 3)  /* Dist to surf > ion_excl_rad */


I'm testing now disabling OCL optimisations.

**Comment by jrprice** [comment]
I've just pushed a fix for this (76b1c25), and tested that I can now run the `gemnoui` test from the `OpenDwarfs` suite without any errors.

I'm not planning on doing a new binary release any time soon, so I've created [an updated Linux build tarball](https://www.dropbox.com/s/4vgp9kga2eo8w7i/Oclgrind-15.5-Linux-76b1c25.tgz?dl=0) so that you don't have to build Oclgrind from source. Let me know if it doesn't work for any reason, or if you're still experiencing the issues that you reported.

Thanks again for reporting the bug!
